lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1250 @@
+  (0x1D000 + ((_idx)*REG_DMA_BD_BLOCK_SIZE) + 0x14)
+#define REG_DMA_CTRL_BD(_idx) (0x1D000 + ((_idx)*REG_DMA_BD_BLOCK_SIZE) + 0x18)
+
+// DMA S2MM channel control
+#define REG_DMA_S2MM_BLOCK_SIZE 0x08
+#define REG_DMA_S2MM_CTRL(_channel)                                            \
+  (0x1DE00 + ((_channel)*REG_DMA_S2MM_BLOCK_SIZE) + 0x0)
+#define REG_DMA_S2MM_QUEUE(_channel)                                           \
+  (0x1DE00 + ((_channel)*REG_DMA_S2MM_BLOCK_SIZE) + 0x4)
+
+// DMA MM2S channel control
+#define REG_DMA_MM2S_BLOCK_SIZE 0x08
+#define REG_DMA_MM2S_CTRL(_channel)                                            \
+  (0x1DE10 + ((_channel)*REG_DMA_MM2S_BLOCK_SIZE) + 0x0)
+#define REG_DMA_MM2S_QUEUE(_channel)                                           \
+  (0x1DE10 + ((_channel)*REG_DMA_MM2S_BLOCK_SIZE) + 0x4)
+
+static constexpr auto REG_SHM_MUX = 0x1f000u; // a mux?
+
+static constexpr auto REG_SS_MASTER_BLOCK_SIZE =
+    0x60; // stream switch master config 0
+static constexpr auto REG_SS_MASTER_CFG_0 =
+    0x3f000u; // stream switch master config 0
+
+static constexpr auto REG_SS_SLAVE_CFG_BLOCK_SIZE =
+    0x168; // stream switch slave config 0
+static constexpr auto REG_SS_SLAVE_CFG_0 =
+    0x3f100u; // stream switch slave config 0
+
+static constexpr auto REG_SS_SLAVE_CFG_SLOT_BLOCK_SIZE =
+    0x3ac; // stream switch slave config 0, slot 0
+static constexpr auto REG_SS_SLAVE_CFG_0_SLOT_0 =
+    0x3f200u; // stream switch slave config 0, slot 0
+
+// 32KB data memory
+static constexpr auto DATA_MEM_OFFSET = 0x00000u;
+static constexpr auto DATA_MEM_SIZE = 0x08000u;
+
+// 16KB program memory
+static constexpr auto PROG_MEM_OFFSET = 0x20000u;
+static constexpr auto PROG_MEM_SIZE = 0x4000u;
+
+static uint8_t sec_name_offset[SEC_IDX_MAX];
+
+static const char *sec_name_str[SEC_IDX_MAX] = {
+    "null",     ".ssmast",   ".ssslve",    ".sspckt",
+    ".sdma.bd", ".shmmux",   ".sdma.ctl",  ".prgm.mem",
+    ".tdma.bd", ".tdma.ctl", "deprecated", ".data.mem"};
+
+static size_t stridx;
+
+/*
+        Holds a sorted list of all writes made to device memory
+        All recorded writes are time/order invariant. This allows sorting to
+   compact the airbin.
+*/
+static std::map<uint64_t, uint32_t> mem_writes;
+
+/*
+ * Tile address format:
+ * --------------------------------------------
+ * |                7 bits  5 bits   18 bits  |
+ * --------------------------------------------
+ * | Array offset | Column | Row | Tile addr  |
+ * --------------------------------------------
+ */
+class TileAddress {
+public:
+  TileAddress(uint8_t column, uint8_t row, uint64_t array_offset = 0x000u)
+      : array_offset{array_offset}, column{column}, row{row} {}
+
+  // SFINAE is used here to choose the copy constructor for `TileAddress`,
+  // and this constructor for all other classes.
+  template <
+      typename Op,
+      std::enable_if_t<not std::is_same<Op, TileAddress>::value, bool> = true>
```

Comment:
```suggestion
      std::enable_if_t<!std::is_same_t<Op, TileAddress>, bool> = true>
```
`_t` is C++14, and I cannot remember the rule that say we should avoid the alias like `not`.
We can shorten a little bit this, but anyway all this will be replaced by `requires` from C++20 when we switch to.

Reply:
done